### PR TITLE
Restrict to proxy controllers to external servers

### PR DESCRIPTION
**Pull request by sureshc**
Prevent Media Proxy and other proxy controllers from being used to communicate to internal servers.

### CODE REVIEW

File: dashboard/app/helpers/proxy_helper.rb [hunk 1]
```diff
@@ -147,4 +160,11 @@ def render_error_response(status, text)
     prevent_caching
     render plain: text, status: status
   end
+
+  # Do not permit proxying to a server on our own private network, unless it is our own dashboard IP Address (we
+  # sometimes proxy to ourselves, which is an internal IP address on development / continuous integration environments).
```
Comment (davidsbailey) [inline dashboard/app/helpers/proxy_helper.rb:160]: do we still need this exception allowing our own dashboard IP address, given the comments elsewhere that this is only enforced in production?

**Review by davidsbailey — COMMENTED** [review]
currently, "code.org" is whitelisted in the xhr proxy controller, and some apps (perhaps some curriculum?) depend on applab apps being able to use this to read data from each other. Will this change break this behavior?

in security, allow-lists are typically preferred over deny lists. is there a way to assert positively that a request is going to some place outside our network rather than trying to list all of the ways that it could be going somewhere inside?

File: dashboard/app/helpers/proxy_helper.rb (hunk 1 above)
Comment (sureshc) [reply dashboard/app/helpers/proxy_helper.rb:160]: I've eliminated the production-only logic in favor of a more narrow exception that allows proxying to a private IP address when that is the IP address that the current dashboard server is operating on.

**Comment by sureshc** [comment]
> currently, "code.org" is whitelisted in the xhr proxy controller, and some apps (perhaps some curriculum?) depend on applab apps being able to use this to read data from each other. Will this change break this behavior?

code.org resolves to a public address, so proxying to it is allowed and not impacted by this change.

> in security, allow-lists are typically preferred over deny lists. is there a way to assert positively that a request is going to some place outside our network rather than trying to list all of the ways that it could be going somewhere inside?

As best as I can tell, the formal [definition of a public IP address](https://en.wikipedia.org/wiki/IP_address#Public_address), is any IP address that is not reserved for private/local use.  I've tried to use a method name that is consistent with the practice of defining allow-lists (`allowed_ip_address?`), but the logic within that method needs to check the inverse, that the IP address is not a private/reserved address as [defined by the IANA](https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml).

**Review by davidsbailey — APPROVED** [review]
thanks for responding with well-thought-through answers! LGTM!

File: dashboard/app/helpers/proxy_helper.rb [hunk 2]
```diff
@@ -5,6 +5,19 @@
...
+  # https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml
+  PRIVATE_IPS = [
+    IPAddr.new('0.0.0.0/8'),
+    IPAddr.new('10.0.0.0/8'),
+    IPAddr.new('127.0.0.0/8'),
+    IPAddr.new('172.16.0.0/12'),
+    IPAddr.new('169.254.0.0/16'),
+    IPAddr.new('192.168.0.0/16'),
+    IPAddr.new('fd00::/8')
+  ].freeze
```
Comment (wjordan) [inline dashboard/app/helpers/proxy_helper.rb]: I think this check could be simplified to `ip.private? || ip.link_local? || ip.loopback? || IPAddr.new('0.0.0.0/8').include?(ip)`, which would offload some of these details to the Ruby standard lib.

**Review by wjordan — APPROVED** [review]
Overall LGTM, see comment about potential simplification.

File: dashboard/app/helpers/proxy_helper.rb (hunk 2 above)
Comment (sureshc) [reply dashboard/app/helpers/proxy_helper.rb]: I didn't notice those handy methods.  Switched to using them!

**Pull request merged**